Added support for Tuple operations in the CFG.

333fred opened the pull request:
Tagging @dotnet/roslyn-compiler @dotnet/analyzer-ioperation @AlekseyTs for review.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs:
```diff
@@ -1337,5 +1337,347 @@ static void Main()
+    void M(bool b)
+    /*<bind>*/{
+        int i1, i2, i3;
+        (i1, b ? i2 : i3) = (1, 2);
```
>(i1, b ? i2 : i3) = (1, 2); [](start = 8, length = 27)

It would be good to test success scenario like ```(i1, b1 ? ref i2 : ref i3) = (1, 2);```. Also consider making ```i2``` and ```i3``` parameters to avoid "Use of unassigned local variable" diagnostics

AlekseyTs commented:
>         public override IOperation VisitDeconstructionAssignment(IDeconstructionAssignmentOperation operation, int? captureIdForResult)

Since you are touching deconstruction scenarios in testing, it might make sense to adjust implementation of this function and expand the testing accordingly

---
Refers to: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:3091 in eacde95. [](commit_id = eacde953c93bb91675c41311cecdf33e1e2e0f62, deletion_comment = False)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs:
```diff
@@ -1337,5 +1337,347 @@ static void Main()
+{
+    void M(bool b)
+    /*<bind>*/{
+        (int, int) t = (1, b ? 2 : 3);
```
>(1, b ? 2 : 3) [](start = 23, length = 14)

It would be good to cover scenarios when control flow is not in the last element.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ITupleExpression.cs:
```diff
@@ -1337,5 +1337,347 @@ static void Main()
+    Statements (0)
+";
+            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics);
+        }
```
>} [](start = 8, length = 1)

Consider adding relevant tests for VB, or adding a PROTOTYPE comment in the right file with references to the tests that should be ported from C#.

333fred commented:
@mavasani @AlekseyTs I dismissed your reviews, as I added IDeconstructionAssignmentOperation support per Aleksey's suggestion. Should be a small additional change for that.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,557 @@
+          Expression: 
+            IDeconstructionAssignmentOperation (OperationKind.DeconstructionAssignment, Type: (System.Int32 i1, System.Int32 i2)) (Syntax: '(i1, i2) =  ...  : (i5, i6)')
+              Left: 
+                IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: (System.Int32 i1, System.Int32 i2), IsImplicit) (Syntax: '(i1, i2)')
```
>IFlowCaptureReferenceOperation: 0 [](start = 16, length = 33)

I think tracing deconstruction assignment through a capture reference which is a tuple, which in turn can have a control flow is going to be too complicated and unexpected because language doesn't allow arbitrary expression as a destination. I think we should preserve ITupleOperation as the target of a deconstruction assignment and substitute its elements instead.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,557 @@
+{
+    void M(bool b, int i1, int i2, int i3, int i4, int i5, int i6)
+    /*<bind>*/{
+        (b ? (i1, i2) : (i3, i4)) = (i5, i6);
```
>(b ? (i1, i2) : (i3, i4)) = (i5, i6); [](start = 8, length = 37)

I don't think this is a deconstruction

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,557 @@
+                ITupleOperation (OperationKind.Tuple, Type: (System.Int32 i1, System.Int32 i2)) (Syntax: '(var i1, var i2)')
+                  NaturalType: (System.Int32 i1, System.Int32 i2)
+                  Elements(2):
+                      IDeclarationExpressionOperation (OperationKind.DeclarationExpression, Type: System.Int32) (Syntax: 'var i1')
```
>IDeclarationExpressionOperation (OperationKind.DeclarationExpression, Type: System.Int32) (Syntax: 'var i1') [](start = 22, length = 108)

This IDeclarationExpressionOperation node doesn't have to be spilled and I think we shouldn't spill it.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,557 @@
+{
+    void M(bool b)
+    /*<bind>*/{
+        (b ? (var i1, var i2) : (var i3, var i4)) = (1, 2);
```
>(b ? (var i1, var i2) : (var i3, var i4)) = (1, 2); [](start = 8, length = 51)

I don't think this is a deconstruction

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,557 @@
+        Statements (1)
+            IFlowCaptureOperation: 0 (OperationKind.FlowCapture, Type: null, IsImplicit) (Syntax: 'var (i1, i2)')
+              Value: 
+                IDeclarationExpressionOperation (OperationKind.DeclarationExpression, Type: (System.Int32 i1, System.Int32 i2)) (Syntax: 'var (i1, i2)')
```
>IDeclarationExpressionOperation [](start = 16, length = 31)

I don't think we should spill this node

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,557 @@
+              Expression: 
+                IDeconstructionAssignmentOperation (OperationKind.DeconstructionAssignment, Type: (System.Int32 i1, System.Int32 i2)) (Syntax: 'var (i1, i2 ... 2) : (3, 4)')
+                  Left: 
+                    IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: (System.Int32 i1, System.Int32 i2), IsImplicit) (Syntax: 'var (i1, i2)')
```
>IFlowCaptureReferenceOperation [](start = 20, length = 30)

I would expect an IDeclarationExpressionOperation node here

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,557 @@
+              Expression: 
+                IDeconstructionAssignmentOperation (OperationKind.DeconstructionAssignment, Type: (System.Int32 i1, System.Int32 i2)) (Syntax: 'var (i1, i2 ... ? this : c1')
+                  Left: 
+                    IFlowCaptureReferenceOperation: 0 (OperationKind.FlowCaptureReference, Type: (System.Int32 i1, System.Int32 i2), IsImplicit) (Syntax: 'var (i1, i2)')
```
>IFlowCaptureReferenceOperation [](start = 20, length = 30)

I would expect IDeclarationExpressionOperation of original shape here.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,557 @@
+{
+    void M(bool b)
+    /*<bind>*/{
+        (var i1, var i2) = b ? (1, 2) : (3, 4);
```
>(var i1, var i2) [](start = 8, length = 16)

Consider also testing a mixed case:
```
int i2;
(int i1, i2) = ...
``` 

This is an error today, but the transformation we do should still make sense.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeclarationExpression.cs:
```diff
@@ -0,0 +1,111 @@
+{
+    void M(bool b)
+    /*<bind>*/{
+        M2(b ? out var i1 : out var i2);
```
> M2(b ? out var i1 : out var i2) [](start = 7, length = 32)

Do we even have a declaration expression in the parse tree for this scenario?

AlekseyTs commented:
Done with review pass (iteration 3).

333fred commented:
@AlekseyTs addressed feedback. I moved all the nondeconstruction test cases out to other files.

---
In reply to: [380953595](https://github.com/dotnet/roslyn/pull/26108#issuecomment-380953595) [](ancestors = 380953595)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2791,6 +2792,50 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            }
+            else
+            {
+                // This is an error case, so we don't try to do any kind of simplification for it
```
>// This is an error case, [](start = 16, length = 25)

I don't think this is accurate, this could be a DeclarationExpression in success scenario.

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2791,6 +2792,50 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+            if (operation.Target.Kind == OperationKind.Tuple)
+            {
+                tupleTarget = (ITupleOperation)operation.Target;
+                PushArray(tupleTarget.Elements);
```
>PushArray(tupleTarget.Elements); [](start = 16, length = 32)

I don't think this is going to properly handle nested case

AlekseyTs commented:
Done with review pass (iteration 4)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [381233616](https://github.com/dotnet/roslyn/pull/26108#issuecomment-381233616) [](ancestors = 381233616)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IDeconstructionAssignmentOperation.cs:
```diff
@@ -0,0 +1,671 @@
+{
+    void M(bool b, int i1, int i2, int i3)
+    /*<bind>*/{
+        (i1, (i2, i3)) = b ? (1, (2, 3)) : (4, (5, 6));
```
>(i1, (i2, i3)) [](start = 8, length = 14)

Could there be interesting nesting cases with DeclarationExpressions, like ```var (y, (a, b), x??z)```? Basically we don't want capture tuples used within a declaration expressions.

333fred replied:
This case actually doesn't produce a deconstruction, it's an invalid operation.

---
In reply to: [181502121](https://github.com/dotnet/roslyn/pull/26108#discussion_r181502121) [](ancestors = 181502121)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2791,6 +2792,67 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+                if (value.Kind == OperationKind.Tuple)
+                {
+                    var tuple = (ITupleOperation)value;
+                    var elementBuilder = ArrayBuilder<IOperation>.GetInstance();
```
>GetInstance() [](start = 66, length = 13)

Do we know the size of the builder that we need?

333fred replied:
We do, I'll update.

---
In reply to: [181503256](https://github.com/dotnet/roslyn/pull/26108#discussion_r181503256) [](ancestors = 181503256)

AlekseyTs commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -2791,6 +2792,67 @@ public override IOperation VisitObjectCreation(IObjectCreationOperation operatio
+
+        public override IOperation VisitDeclarationExpression(IDeclarationExpressionOperation operation, int? captureIdForResult)
+        {
+            return new DeclarationExpression(Visit(operation.Expression), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);
```
>return new DeclarationExpression(Visit(operation.Expression), semanticModel: null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit); [](start = 12, length = 164)

Consider reusing logic from ```VisitDeconstructionAssignment``` if ```operation.Expression``` is a tuple. I think this will make implementation more robust. We want to guarantee that the shape of the structure is not changed no matter what.

333fred replied:
I'll do this in a followup PR, filed [https://github.com/dotnet/roslyn/issues/26151](https://github.com/dotnet/roslyn/issues/26151) to track.

---
In reply to: [181504915](https://github.com/dotnet/roslyn/pull/26108#discussion_r181504915) [](ancestors = 181504915)

333fred commented:
Ping @mavasani @dotnet/analyzer-ioperation @dotnet/roslyn-compiler for reviews please.

jcouv commented on src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs:
```diff
@@ -1266,11 +1266,12 @@ private void SpillEvalStack()
                 {
                     int captureId = _availableCaptureId++;
 
-                    AddStatement(new FlowCapture(captureId, operation.Syntax, Visit(operation))
```
I'm surprised by the change here. We don't need to visit `operation`?

333fred replied:
Everything that is on `_evalStack` has already been visited. This extra call to Visit is actually a bug that was causing issues with DeconstructionFlow_07.

---
In reply to: [181832673](https://github.com/dotnet/roslyn/pull/26108#discussion_r181832673) [](ancestors = 181832673)

333fred commented:
Manually squashing commits before merge.
